src/main/java/com/github/adeynack/executors/LimitedExecutor.java

```diff
@@ -0,0 +1,92 @@
+package com.github.adeynack.executors;
+
+import org.jetbrains.annotations.NotNull;
+
+import javax.annotation.concurrent.GuardedBy;
+import java.util.ArrayDeque;
+import java.util.Queue;
+import java.util.concurrent.Executor;
+import java.util.concurrent.locks.ReentrantLock;
+
+@SuppressWarnings("WeakerAccess") // Library class
+public class LimitedExecutor implements Executor {
+
+  private final ReentrantLock mainLock = new ReentrantLock();
+
+  private final Executor baseExecutor;
+
+  private final int parallelTaskLimit;
+
+  @GuardedBy("mainLock")
+  private final Queue<Runnable> taskQueue = new ArrayDeque<>();
+
+  /**
+   * indicates the number of tasks already submitted to the underlying {@link Executor}.
+   */
+  @GuardedBy("mainLock")
+  private volatile int tasksSubmitted = 0;
+
+  public LimitedExecutor(Executor baseExecutor, int parallelTaskLimit) {
+    this.baseExecutor = baseExecutor;
+    this.parallelTaskLimit = parallelTaskLimit;
+  }
+
+  @Override
+  public void execute(@NotNull Runnable command) {
+    mainLock.lock();
+    try {
+      if (tasksSubmitted >= parallelTaskLimit) {
+        taskQueue.add(command);
+      } else {
+        final int actualTasksSubmitted = tasksSubmitted;
+        tasksSubmitted = actualTasksSubmitted + 1;
```

Comment:
why not `tasksSubmitted++` instead of 2-line dance
```suggestion
        tasksSubmitted++;
```